syntax = "proto3";

message Plane {
    float normal_x = 1;
    float normal_y = 2;
    float normal_z = 3;
    float distance = 4;
}

message TextureMatrix {
    float u_offset_x = 1;
    float u_offset_y = 2;
    float u_offset_z = 3;

    float v_offset_x = 4;
    float v_offset_y = 5;
    float v_offset_z = 6;
}

message PrimitiveBrushSide {
    string material = 1;
    Plane plane = 2;
    TextureMatrix texture_matrix = 3;
}

message PrimitiveBrush {
    repeated PrimitiveBrushSide sides = 1;
}

message PrimitivePatch {
    string material = 1;
    bool explicitly_subdivided = 2;
    int32 subdivisions_horizontal = 3;
    int32 subdivisions_vertical = 4;
}

message Primitive {
    map<string, string> configuration = 1;

    oneof primitive_type {
        PrimitiveBrush brush = 2;
        PrimitivePatch patch = 3;
    }
}

message Entity {
    map<string, string> configuration = 1;

    repeated Primitive primitives = 2;
}

message Map {
    string name = 1;
    repeated Entity entities = 2;
}
